Feature/disable-offer-visibility

Naapperas opened the pull request:
Closes #218

TiagooGomess reviewed (commented):
Change the name of the state for loadingOfferVisibility

TiagooGomess reviewed (changes requested):
I suggest making this change so that the color of the visibility button when the company or offer are blocked also changes:

![image](https://user-images.githubusercontent.com/50210213/162751853-fdcb99da-391a-4c40-af3d-a013b310a955.png)

TiagooGomess reviewed (commented):
On the offer page, the visibility button does not become disabled while the request is being fulfilled. Can you replicate that on your machine?

Naapperas commented:
While testing I left a `console.log` in the OfferVIsibilityOptions file and both the "offer details" on the homepage search section and the offer page itself, when pressing the "hid/enable" button, triggered that log, so I assumed that it worked for both, but i might be mistaken.

TiagooGomess commented:
> While testing I left a `console.log` in the OfferVIsibilityOptions file and both the "offer details" on the homepage search section and the offer page itself, when pressing the "hide/enable" button, triggered that log, so I assumed that it worked for both, but i might be mistaken.

You should enable throttling on the dev tools network tab (for example, slow 3G) so that you can simulate a request that takes some time to be fulfilled. Doing that, can you see the visibility button being disabled?

TiagooGomess commented:
> Also, should i write tests covering these scenarios ?

You should look at the coverage report and see if there are some uncovered lines (that cover added functionality).

Naapperas commented:
> > While testing I left a `console.log` in the OfferVIsibilityOptions file and both the "offer details" on the homepage search section and the offer page itself, when pressing the "hide/enable" button, triggered that log, so I assumed that it worked for both, but i might be mistaken.
> 
> You should enable throttling on the dev tools network tab (for example, slow 3G) so that you can simulate a request that takes some time to be fulfilled. Doing that, can you see the visibility button being disabled?

By explicitly waiting 2 seconds (setTimeout) the button appears to be disabled, like this 

![image](https://user-images.githubusercontent.com/23704084/162792388-23316e31-1a4d-41b3-a29e-902c7630da78.png)

and it does not fire any events (thus not performing any request to the backend), so we know that the button is disabled at some point. 

What I think is happening is the disabling/enabling is happening faster than the request itself (enabling GPRS or normal 2G, which results in slow requests, didn't seem to make the button get disabled). I added this behavior in 

```js
    const handleOfferVisibility = async () => {
        setLoadingOfferVisibility(true);
        if (visibilityState.isVisible) {
            await handleHideOffer({
                offer: offer,
                addSnackbar: addSnackbar,
                onError: onError,
            });
        } else {
            await handleCompanyEnableOffer({
                offer: offer,
                addSnackbar: addSnackbar,
                onError: onError,
            });
        }
        setLoadingOfferVisibility(false);
    };
```

thinking that the "enabling" part" would only occur after the requests awaited, but it does not seem to be the case. Maybe I could remove the await statement and write this as a then/catch chain, what do you think ?

TiagooGomess commented:
> > > While testing I left a `console.log` in the OfferVIsibilityOptions file and both the "offer details" on the homepage search section and the offer page itself, when pressing the "hide/enable" button, triggered that log, so I assumed that it worked for both, but i might be mistaken.
> > 
> > 
> > You should enable throttling on the dev tools network tab (for example, slow 3G) so that you can simulate a request that takes some time to be fulfilled. Doing that, can you see the visibility button being disabled?
> 
> By explicitly waiting 2 seconds (setTimeout) the button appears to be disabled, like this
> 
> ![image](https://user-images.githubusercontent.com/23704084/162792388-23316e31-1a4d-41b3-a29e-902c7630da78.png)
> 
> and it does not fire any events (thus not performing any request to the backend), so we know that the button is disabled at some point.
> 
> What I think is happening is the disabling/enabling is happening faster than the request itself (enabling GPRS or normal 2G, which results in slow requests, didn't seem to make the button get disabled). I added this behavior in
> 
> ```js
>     const handleOfferVisibility = async () => {
>         setLoadingOfferVisibility(true);
>         if (visibilityState.isVisible) {
>             await handleHideOffer({
>                 offer: offer,
>                 addSnackbar: addSnackbar,
>                 onError: onError,
>             });
>         } else {
>             await handleCompanyEnableOffer({
>                 offer: offer,
>                 addSnackbar: addSnackbar,
>                 onError: onError,
>             });
>         }
>         setLoadingOfferVisibility(false);
>     };
> ```
> 
> thinking that the "enabling" part" would only occur after the requests awaited, but it does not seem to be the case. Maybe I could remove the await statement and write this as a then/catch chain, what do you think ?

Yes, you can try that

TiagooGomess commented:
Not working correctly on my machine:

https://user-images.githubusercontent.com/50210213/162983422-a139dadf-b069-401f-85d0-01cdc3bb1938.mov

TiagooGomess commented:
When the offer is hidden, the color of the visibility button on the company management offers page is "undefined". It should be secondary. And when the offer is disabled, the color is secondary, but it should be "undefined".

TiagooGomess commented on src/components/Company/Offers/Manage/CompanyOffersVisibilityActions.js:
```diff
@@ -99,15 +104,30 @@ const CompanyOffersVisibilityActions = ({ offer }) => {
-                {offerVisibilityState.isVisible ? <VisibilityOffIcon data-testid="HideOffer" color="secondary" fontSize="medium" />
-                    : <VisibilityIcon data-testid="EnableOffer" color="secondary" fontSize="medium" />}
-            </IconButton>
+            <>
```
```suggestion
            
```

TiagooGomess commented on src/components/Company/Offers/Manage/CompanyOffersVisibilityActions.js:
```diff
@@ -99,15 +104,30 @@ const CompanyOffersVisibilityActions = ({ offer }) => {
+                        />
+                    }
+                </IconButton>
+            </>
```
```suggestion
            
```

TiagooGomess commented on src/components/HomePage/SearchResultsArea/Offer/OfferVisibilityOptions.js:
```diff
@@ -75,31 +73,33 @@ const OfferVisibilityOptions = ({
-                                { visibilityState.isVisible ? "Hide Offer" : "Enable Offer" }
-                            </Button>
+                        <Tooltip title={visibilityState.isVisible ? "Hide Offer" : "Enable Offer"}>
+                            <>
```
```suggestion
                           
```

TiagooGomess commented on src/components/HomePage/SearchResultsArea/Offer/OfferVisibilityOptions.js:
```diff
@@ -75,31 +73,33 @@ const OfferVisibilityOptions = ({
+                                >
+                                    {visibilityState.isVisible ? "Hide Offer" : "Enable Offer"}
+                                </Button>
+                            </>
```
```suggestion
                            
```

TiagooGomess reviewed (changes requested):
You must remove these; otherwise, the tooltips don't work.

Naapperas commented:
> You must remove these; otherwise, the tooltips don't work.

I had changed those because MUI would complain about providing a disabled child to the Tooltip component and I figured that a React fragment would work, changed it to a `span` instead.

TiagooGomess commented:
Do the same for [OfferVisibilityOptions.js](https://github.com/NIAEFEUP/nijobs-fe/pull/233/files/612dd757c7f466c9640a4a38af9d3e40ed44a67a#diff-5004fe71a0fb5f48ecd3f9cbe5301cce777f84e76a9105d0c0e62ffc81127283)

TiagooGomess commented on src/services/locationSearchService.js:
```diff
@@ -1,27 +1,37 @@
 export const searchCities = async (prefix, limit = 10) => {
 
-    const API_ENDPOINT = "https://wft-geo-db.p.rapidapi.com/v1/geo/cities";
+    // const API_ENDPOINT = "https://wft-geo-db.p.rapidapi.com/v1/geo/cities";
     const options = {
-        types: "CITY",
-        minPopulation: "100000",
-        namePrefix: prefix,
+        // types: "CITY",
+        // minPopulation: "100000",
+        // namePrefix: prefix,
         limit,
+        searchTerm: encodeURI(prefix),
     };
+
     try {
-        const res = await fetch(`${API_ENDPOINT}?${Object.entries(options).map(([k, v]) => `${k}=${v}`).join("&")}`, {
+        const res = await fetch(`${API_HOSTNAME}/location/search?${Object.entries(options).map(([k, v]) => `${k}=${v}`).join("&")}`, {
             method: "GET",
             headers: {
-                "x-rapidapi-host": "wft-geo-db.p.rapidapi.com",
-                "x-rapidapi-key": process.env.GEO_API_KEY,
+                // "x-rapidapi-host": "wft-geo-db.p.rapidapi.com",
+                // "x-rapidapi-key": process.env.GEO_API_KEY,
             },
         });
 
         if (!res.ok) return [];
 
-        const data = (await res.json()).data;
-        return data.map(({ city, country, latitude, longitude }) => ({ city, country, latitude, longitude }));
+        // const data = (await res.json()).data;
+        // return data.map(({ city, country, latitude, longitude }) => ({ city, country, latitude, longitude }));
+
+        const data = await res.json();
+
+        return data;
 
     } catch (err) {
         throw [{ msg: Constants.UNEXPECTED_ERROR_MESSAGE }];
```
Maybe you should not modify this file in this PR

Naapperas replied:
I don't now how this got through...
